src/engine/export/mmlOps.cpp

```diff
@@ -0,0 +1,427 @@
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "../fileOps/fileOpsCommon.h"
+#include "mmlHelpers.h"
+
+SafeWriter* DivEngine::saveMMLGB(bool useLegacyNoiseTable) {
+  stop();
+  repeatPattern = false;
+  shallStop = false;
+  setOrder(0);
+
+  BUSY_BEGIN_SOFT;
+
+  // Determine loop point in the song (order, row, loopEnd)
+  int loopOrder = 0, loopRow = 0, loopEnd = 0;
+  walkSong(loopOrder, loopRow, loopEnd);
+  logI("loop point: %d %d", loopOrder, loopRow);
+  bool loopMarkerWritten[4] = {false, false, false, false};
+
+  SafeWriter* w = new SafeWriter;
+  w->init();
+
+  // Write header info with song and subsong metadata
+  w->writeText("; Furnace MML-GB Output\n;\n");
+  w->writeText("; Information:\n");
+  w->writeText(fmt::sprintf("; \tname: %s\n", song.name));
+  w->writeText(fmt::sprintf("; \tauthor: %s\n", song.author));
+  w->writeText(fmt::sprintf("; \tcategory: %s\n", song.category));
+  w->writeText(";\n");
+  w->writeText("; SubSongInformation:\n");
+  w->writeText(fmt::sprintf("; \tname: %s\n", curSubSong->name));
+  w->writeText("\n");
+
+  // Write WaveTable Macros, default or custom
+  w->writeText("; WaveTable Macros:\n");
+  if (song.waveLen == 0) {
+    w->writeText("@wave0 = {");
+    _writeDefaultWave(w);
+    w->writeText("}\n");
+  } else {
+    for (int i = 0; i < song.waveLen; i++) {
+      auto wave = getWave(i);
+      w->writeText(fmt::sprintf("@wave%d = {", i));
+      _writeNormalizedGBWave(wave, w);
+      w->writeText("}\n");
+    }
+  }
+  w->writeText("\n");
+  w->writeText("; Sequence data:\n");
+
+  // Prepare for playback and MML generation
+  playSub(false);
+  bool done = false;
+  int tick = 0;
+  bool oldCmdStreamEnabled = cmdStreamEnabled;
+  cmdStreamEnabled = true;
+  double curDivider = divider;
+
+  // Strings to accumulate MML output per channel
+  String mmlChanStream[4] = {"", "", "", ""};
+  String chanNames[4] = {"A", "B", "C", "D"};
+
+  // Rows per measure from current subsong highlighting
+  int rowsPerMeasure = curSubSong->hilightB;
+
+  // Calculate tempo for MML export based on current subsong and divider
+  int tempo = _computeMmlTempo(curSubSong, curDivider);
+
+  // Write initial tempo commands for all channels
+  for (const auto& cname : chanNames) {
+    w->writeText(fmt::sprintf("%s t%d\n", cname.c_str(), tempo));
+  }
+  w->writeText("\n\n");
+
+  // MML generation state
+  _MMLGBState st{};
+  st.currTempo = tempo;
+  st.tempoTick = -1;
+  std::fill(std::begin(st.cmdTick), std::end(st.cmdTick), 0);
+  std::fill(std::begin(st.prevOrder), std::end(st.prevOrder), -1);
+  std::fill(std::begin(st.prevRow), std::end(st.prevRow), -1);
+  std::fill(std::begin(st.chanNameUsed), std::end(st.chanNameUsed), false);
+  std::fill(std::begin(st.noteOn), std::end(st.noteOn), false);
+
+  // Helper to update the MML stream of a channel given elapsed ticks
+  auto updateMmlStream = [&](int chan, int ticksElapsed) {
+    auto& mmlStream = mmlChanStream[chan];
```

Comment:
Don't auto here.
(Don't worry about the `auto` in the lambda definition - it is required)